src/org/exist/security/PermissionFactory.java

```diff
@@ -110,40 +111,49 @@ public static Permission getPermission(final SecurityManager sm, final String us
     }
 
     public static void updatePermissions(final DBBroker broker, final XmldbURI pathUri, final ConsumerE<Permission, PermissionDeniedException> permissionModifier) throws PermissionDeniedException {
-        DocumentImpl doc = null;
         final TransactionManager transact = broker.getBrokerPool().getTransactionManager();
         try(final Txn transaction = transact.beginTransaction()) {
-            final Collection collection = broker.openCollection(pathUri, LockMode.WRITE_LOCK);
-            if (collection == null) {
-                doc = broker.getXMLResource(pathUri, LockMode.WRITE_LOCK);
-                if(doc == null) {
-                    transact.abort(transaction);
-                    throw new XPathException("Resource or collection '" + pathUri.toString() + "' does not exist.");
-                }
+            Collection collection = null;
+            try {
+                collection = broker.openCollection(pathUri, LockMode.WRITE_LOCK);
+                if (collection == null) {
+                    DocumentImpl doc = null;
+                    try {
+                        doc = broker.getXMLResource(pathUri, LockMode.WRITE_LOCK);
+                        if (doc == null) {
+                            transact.abort(transaction);
+                            throw new XPathException("Resource or collection '" + pathUri.toString() + "' does not exist.");
+                        }
 
-                final Permission permissions = doc.getPermissions();
-                permissionModifier.accept(permissions);
+                        final Permission permissions = doc.getPermissions();
+                        permissionModifier.accept(permissions);
 
-                broker.storeXMLResource(transaction, doc);
-                transact.commit(transaction);
-                broker.flush();
-            } else {
-                // keep the write lock in the transaction
-                transaction.registerLock(collection.getLock(), LockMode.WRITE_LOCK);
+                        broker.storeXMLResource(transaction, doc);
+                    } finally {
+                        if(doc != null) {
+                            doc.getUpdateLock().release(LockMode.WRITE_LOCK);
+                        }
+                    }
+                    transact.commit(transaction);
```

Comment:
Considering that Transactions in eXist are not proper transactions; Can you explain why?